internal/terraform/context_apply_deferred_test.go

```diff
@@ -1574,8 +1580,393 @@ output "a" {
+					"a": cty.ObjectVal(map[string]cty.Value{
+						"name":           cty.StringVal("old_value"),
+						"upstream_names": cty.NullVal(cty.Set(cty.String)),
+						"output":         cty.StringVal("computed_output"),
+					}),
+				},
+				wantDeferred: map[string]ExpectedDeferred{
+					"test.a": {Reason: providers.DeferredReasonProviderConfigUnknown, Action: plans.DeleteThenCreate},
+				},
+				complete: false,
+			},
+		},
+	}
+
+	// planDeleteResourceChange is a test that covers the behavior of planning a resource that is removed from the config.
+	planDeleteResourceChange = deferredActionsTest{
+		configs: map[string]string{
+			"main.tf": `
+// Empty config, expect to delete everything
+		`,
+		},
+		state: states.BuildState(func(state *states.SyncState) {
+			state.SetResourceInstanceCurrent(
+				mustResourceInstanceAddr("test.a"),
+				&states.ResourceInstanceObjectSrc{
+					Status: states.ObjectReady,
+					AttrsJSON: mustParseJson(map[string]interface{}{
+						"name":   "deferred_resource_change",
+						"output": "computed_output",
+					}),
+				},
+				addrs.AbsProviderConfig{
+					Provider: addrs.NewDefaultProvider("test"),
+					Module:   addrs.RootModule,
+				})
+		}),
+		stages: []deferredActionsTestStage{
+			{
+
+				inputs:      map[string]cty.Value{},
+				wantPlanned: map[string]cty.Value{},
+
+				wantActions: map[string]plans.Action{},
+				wantApplied: map[string]cty.Value{
+					// The all resources will be deferred, so shouldn't
+					// have any action at this stage.
+				},
+				wantOutputs: map[string]cty.Value{},
+				wantDeferred: map[string]ExpectedDeferred{
+					"test.a": {Reason: providers.DeferredReasonProviderConfigUnknown, Action: plans.Delete},
+				},
+				complete: false,
+			},
+		},
+	}
+
+	// planDestroyResourceChange is a test that covers the behavior of planning a resource
+	planDestroyResourceChange = deferredActionsTest{
+		configs: map[string]string{
+			"main.tf": `
+resource "test" "a" {
+	name = "deferred_resource_change"
+}
+output "a" {
+	value = test.a
+}
+		`,
+		},
+		state: states.BuildState(func(state *states.SyncState) {
+			state.SetResourceInstanceCurrent(
+				mustResourceInstanceAddr("test.a"),
+				&states.ResourceInstanceObjectSrc{
+					Status: states.ObjectReady,
+					AttrsJSON: mustParseJson(map[string]interface{}{
+						"name": "deferred_resource_change",
+					}),
+				},
+				addrs.AbsProviderConfig{
+					Provider: addrs.NewDefaultProvider("test"),
+					Module:   addrs.RootModule,
+				})
+		}),
+		stages: []deferredActionsTestStage{
+			{
+				buildOpts: func(opts *PlanOpts) {
+					opts.Mode = plans.DestroyMode
+				},
+				inputs: map[string]cty.Value{},
+				wantPlanned: map[string]cty.Value{
+					"deferred_resource_change": cty.ObjectVal(map[string]cty.Value{
+						"name":           cty.StringVal("deferred_resource_change"),
+						"upstream_names": cty.NullVal(cty.Set(cty.String)),
+						"output":         cty.UnknownVal(cty.String),
+					}),
```

Comment:
I think something isn't right here? The previous destroy test which has an empty `wantPlanned` - I'd expect the same thing to happen here. It looks like it's getting the action right, but seems to be passing in proposed state to the provider when it shouldn't. Do you have any idea why that might be happening?

Reply:
I am a bit puzzled as well, I compared the two test cases. To me it makes sense that there is sth in wantPlanned since we run the plan and the plan is deferred. [one goes the node_resource_destroy_deposed route](https://github.com/hashicorp/terraform/blob/ea9162235da499bf33c010b3b1020642cf768ae2/internal/terraform/node_resource_destroy_deposed.go#L176) and the other the [node_resource_plan_orphan route](https://github.com/hashicorp/terraform/blob/ea9162235da499bf33c010b3b1020642cf768ae2/internal/terraform/node_resource_plan_orphan.go#L177) but in both cases the same abstract plan destroy is called. No idea how to find out where the difference originates from, do you have a rough idea where the difference could be?